xbmc/addons/VFSEntry.cpp

```diff
@@ -281,47 +330,62 @@ int64_t CVFSEntry::GetLength(void* ctx)
 
 int CVFSEntry::IoControl(void* ctx, XFILE::EIoControl request, void* param)
 {
-  if (!m_struct.toAddon->io_control)
-    return -1;
-
-  VFS_IOCTRL ctrl = TranslateIOCTRLToAddon(request);
-  if (ctrl == VFS_IOCTRL_INVALID)
-    return -1;
-
-  /*! @note @ref VFS_IOCTRL_NATIVE a call to addon to give data! */
-  if (ctrl == VFS_IOCTRL_NATIVE)
+  switch (request)
   {
-    XFILE::SNativeIoControl* kodiData = static_cast<XFILE::SNativeIoControl*>(param);
-    if (!kodiData)
-      return -1;
-
-    VFS_IOCTRL_NATIVE_DATA data;
-    data.request = kodiData->request;
-    data.param = kodiData->param;
-    return m_struct.toAddon->io_control(&m_struct, ctx, ctrl, &data);
-  }
+    case XFILE::EIoControl::IOCTRL_SEEK_POSSIBLE:
+    {
+      if (!m_struct.toAddon->io_control_get_seek_possible)
+        return -1;
+      return m_struct.toAddon->io_control_get_seek_possible(&m_struct, ctx) ? 1 : 0;
+    }
+    case XFILE::EIoControl::IOCTRL_CACHE_STATUS:
+    {
+      if (!m_struct.toAddon->io_control_get_cache_status)
+        return -1;
 
-  /*! @note @ref VFS_IOCTRL_CACHE_STATUS a call to addon to become data from him! */
-  if (ctrl == VFS_IOCTRL_CACHE_STATUS)
-  {
-    XFILE::SCacheStatus* kodiData = static_cast<XFILE::SCacheStatus*>(param);
-    if (!kodiData)
-      return -1;
+      XFILE::SCacheStatus* kodiData = static_cast<XFILE::SCacheStatus*>(param);
+      if (!kodiData)
+        return -1;
 
-    VFS_IOCTRL_CACHE_STATUS_DATA data = {0};
-    int ret = m_struct.toAddon->io_control(&m_struct, ctx, ctrl, &data);
-    if (ret >= 0)
+      VFS_CACHE_STATUS_DATA status;
+      int ret = m_struct.toAddon->io_control_get_cache_status(&m_struct, ctx, &status) ? 0 : -1;
+      if (ret >= 0)
+      {
+        kodiData->forward = status.forward;
+        kodiData->maxrate = status.maxrate;
+        kodiData->currate = status.currate;
+        kodiData->lowspeed = status.lowspeed;
+      }
+      return ret;
+    }
+    case XFILE::EIoControl::IOCTRL_CACHE_SETRATE:
     {
-      kodiData->forward = data.forward;
-      kodiData->maxrate = data.maxrate;
-      kodiData->currate = data.currate;
-      kodiData->lowspeed = data.lowspeed;
+      if (!m_struct.toAddon->io_control_set_cache_rate)
+        return -1;
+
+      return m_struct.toAddon->io_control_set_cache_rate(&m_struct, ctx,
+                                                         *static_cast<unsigned int*>(param))
+                 ? 1
+                 : 0;
```

Comment:
The clang formatting of this line is a little awkward. If you alias `param` you could probably fit everything on one line:

```c++
unsigned int& iParam = *static_cast<unsigned int*>(param);
```

Reply:
Thanks, with them looks much cleaner, have changed to them.